### PR TITLE
Implement Condition Variables in tokio-util

### PR DESCRIPTION
**Pull request by kaimast**
## Motivation
See #2739. 

## Solution
This adds a Condition Variable implementation on top of `tokio::sync::Notifier` to `tokio-util`.

## Note
This pull request relies on #3741 to undo the modifications to `Cargo.toml`.

### CODE REVIEW

File: tokio/src/sync/mutex.rs [hunk 1]
```diff
@@ -561,6 +561,12 @@ impl<'a, T: ?Sized> MutexGuard<'a, T> {
             marker: marker::PhantomData,
         })
     }
+
+    /// Releases the lock and returns a reference to the corresponding Mutex
+    pub fn into_mutex_reference(self) -> &'a Mutex<T> {
+        self.lock.s.release(1);
+        self.lock
+    }
```
Comment (Darksonn) [inline tokio/src/sync/mutex.rs:569]: I'm not sure that exposing this method is great.

File: tokio-util/src/sync/condvar.rs [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+    /// Notify at most one other thread
+    pub fn notify_one(&self) {
+        let count = self.wait_count.load(atomic::Ordering::SeqCst);
+
+        if count > 0 {
+            self.wait_count.store(count-1, atomic::Ordering::SeqCst);
+            self.notifier.notify_one();
+        }
+    }
```
Comment (Darksonn) [inline tokio-util/src/sync/condvar.rs:34]: This has a race condition if multiple threads call `notify_one` concurrently. You would need to do a compare and swap in a loop or similar.

File: tokio-util/src/sync/condvar.rs [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+
+    /// Notify all waiting threads
+    pub fn notify_all(&self) {
+        let count = self.wait_count.swap(0, atomic::Ordering::SeqCst);
+
+        for _ in 0..count {
+            self.notifier.notify_one();
+        }
```
Comment (Darksonn) [inline tokio-util/src/sync/condvar.rs:42]: The `Notify` utility has a method for notifying everyone.

**Review by Darksonn — COMMENTED** [review]
It seems to me that if we have a `Condvar`, it should be next to the mutex rather than in `tokio-util`. If nothing else to avoid exposing that `into_mutex_reference` method.

**Comment by kaimast** [comment]
> It seems to me that if we have a `Condvar`, it should be next to the mutex rather than in `tokio-util`. If nothing else to avoid exposing that `into_mutex_reference` method.

That makes sense. Not sure what the process is to get new features into tokio proper?

File: tokio-util/src/sync/condvar.rs (hunk 3 above)
Comment (kaimast) [reply tokio-util/src/sync/condvar.rs]: Notify's notify_waiters does not tell me how many waiters were notified. So this loop ensure that the number of notified waiters and count stay's in sync.

There might be a cleaner way to implement this though if we made Condvar part of `tokio::sync`.

File: tokio/src/sync/mutex.rs [hunk 4]
```diff
@@ -711,3 +721,48 @@ impl<'a, T: ?Sized + fmt::Display> fmt::Display for MappedMutexGuard<'a, T> {
...
+    /// For for another thread to call notify or notify_all
+    pub async fn wait<'a, T>(&self, lock: MutexGuard<'a, T>) -> MutexGuard<'a, T> {
+        self.wait_count.fetch_add(1, Ordering::SeqCst);
+
+        lock.lock.s.release(1);
+        let mutex = lock.lock;
+
+        self.notifier.notified().await;
+        mutex.lock().await
+    }
```
Comment (Darksonn) [inline tokio/src/sync/mutex.rs:746]: What if someone cancels the `wait` call with a `tokio::select!`?

File: tokio/src/sync/mutex.rs [hunk 5]
```diff
@@ -128,6 +131,13 @@ pub struct Mutex<T: ?Sized> {
     c: UnsafeCell<T>,
 }
 
+/// Behaves like std::sync::Condvar but for tokio Mutexes
+#[ derive(Debug) ]
+pub struct Condvar {
+    notifier: Notify,
+    wait_count: AtomicU32,
```
Comment (Darksonn) [inline tokio/src/sync/mutex.rs:138]: I wonder if we can do without the wait counter entirely.

**Comment by Darksonn** [comment]
We discussed this on discord. It seems that implementing an async `Condvar` that behaves correctly in the face of cancellation is difficult. For example, consider the following situation:
```
Task 1 calls wait.
Task 2 calls wait.
Task 3 calls notify.
Task 1 begins to try to lock the mutex.
Task 1 is cancelled.
```
Here, you would want task 2 to get woken up by the notification since it was ultimately not delivered to task 1. But then consider this:
```
Task 1 calls wait.
Task 2 calls notify.
Task 1 begins to try to lock the mutex.
Task 2 calls wait.
Task 1 is cancelled.
```
In this case, we probably would not want task 2's own notification to wake up task 2. However these two situations require different behavior from task 1 when task 1 gets cancelled.

I would want to see a written down in text proposal for how to tackle this before I read an implementation of it. Some ideas: Maybe it could be handled by allowing spurious wakeups? Maybe it could be handled by requiring the mutex to be locked when calling `notify`?

**Comment by kaimast** [comment]
Thanks for all the feedback, @Darksonn!

I'm still quite new to tokio/async rust and haven't thought about cancellation at all....  Initially the implementation I was using took a mutex guard as argument for the notify function, but I was hoping we could keep the API indentical to that of the standard library.

I'm going to think about it some more, and maybe look at how async-std implements it. I will leave the pull request closed for now and reopen as soon as I have something written up.

**Comment by Darksonn** [comment]
The implementation in async-std just ignores the cancellation question.

**Pull request closed without merging**